tests/contrib/gp/test_models.py

```diff
@@ -71,29 +98,195 @@ def test_model_forward(model_class, X, y, kernel, likelihood):
     loc0, cov0 = gp(Xnew, full_cov=True)
     loc1, var1 = gp(Xnew, full_cov=False)
     assert loc0.dim() == y.dim()
-    assert loc0.size(0) == Xnew.size(0)
-    assert loc0.size()[1:] == y.size()[1:]  # test latent shape
-    assert cov0.dim() == 2
-    assert cov0.size(0) == cov0.size(1)
-    assert cov0.size(0) == Xnew.size(0)
+    assert loc0.shape[-1] == Xnew.shape[0]
+    # test latent shape
+    assert loc0.shape[:-1] == y.shape[:-1]
+    assert cov0.shape[:-2] == y.shape[:-1]
+    assert cov0.shape[-1] == cov0.shape[-2]
+    assert cov0.shape[-1] == Xnew.shape[0]
     assert_equal(loc0, loc1)
-    assert_equal(cov0.diag(), var1)
+    n = Xnew.shape[0]
+    cov0_diag = torch.stack([mat.diag() for mat in cov0.view(-1, n, n)]).reshape(var1.shape)
+    assert_equal(cov0_diag, var1)
 
-    # test trivial forward
-    # for variational models, inferences depend on variational parameters, so skip
+    # test trivial forward: Xnew = X
+    loc, cov = gp(X, full_cov=True)
     if model_class is VariationalGP or model_class is SparseVariationalGP:
-        pass
+        assert_equal(loc.norm().item(), 0)
+        assert_equal(cov, torch.eye(cov.shape[-1]).expand(cov.shape))
     else:
-        loc, cov = gp(X, full_cov=True)
         assert_equal(loc, y)
-        assert_equal(cov.abs().sum().item(), 0)
+        assert_equal(cov.norm().item(), 0)
+
+    # test same input forward: Xnew[0,:] = Xnew[1,:] = ...
+    Xnew = torch.tensor([[2, 3, 1]]).expand(10, 3)
+    loc, cov = gp(Xnew, full_cov=True)
+    loc_diff = loc - loc[..., :1].expand(y.shape[:-1] + (10,))
+    assert_equal(loc_diff.norm().item(), 0)
+    cov_diff = cov - cov[..., :1, :1].expand(y.shape[:-1] + (10, 10))
+    assert_equal(cov_diff.norm().item(), 0)
+
+    # test noise kernel forward: kernel = WhiteNoise
+    gp.kernel = WhiteNoise(input_dim=3, variance=torch.tensor(10))
+    Xnew = torch.tensor([[2, 3, 1]])
+    loc, cov = gp(X, full_cov=True)
+    assert_equal(loc.norm().item(), 0)
+    assert_equal(cov, torch.eye(cov.shape[-1]).expand(cov.shape) * 10)
```

Comment:
Here are forward tests. There are three cases which are easy to calculate: Xnew = X, Xnew = constant, kernel = WhiteNoise.